### DISCUSSION TITLE
MOD and related file support

### DISCUSSION THREAD

**Discussion by madbrain76** [music-assistant/support#2482] [Feature requests and ideas]
MOD and related "tracker" module music files originated in the late 1990s, and were very popular in the 1990s. 
Conceptually, they include PCM samples for various sounds, and some track data, such as notes, and so on.  The player then processes and mixes all this into a stereo PCM stream. A music provider in MA could support these module files.

While new MOD music is not being published every day, there is still some new activity. You can for example see recent uploads and activity at modarchive. .
https://modarchive.org/index.php?request=view_actions_uploads

These formats include :

1. MOD (Amiga): The original tracker format created by Karsten Obarski for the Amiga computer. Commonly used by software like ProTracker, SoundTracker, and NoiseTracker.
2. XM (Extended Module): Developed by FastTracker II, the XM format supports more advanced features compared to the original MOD format, such as more channels and enhanced sample handling.
3. S3M (Scream Tracker 3 Module): Created for the Scream Tracker software, this format introduced features like patterns with variable lengths and additional sample effects.
4. IT (Impulse Tracker Module): Designed for Impulse Tracker, the IT format includes advanced features like digital effects, MIDI support, and more intricate pattern commands.
5. MTM (MultiTracker Module): A less common format created for the MultiTracker software, similar to MOD but with support for more channels and other enhancements.
6. STM (Scream Tracker 2 Module): An earlier format used by Scream Tracker 2, preceding the more advanced S3M format.
7. 669 (Composer 669 Module): An early format used by the 669 Composer and UNIS 669, known for its simplicity and limited channel support.
8. MED (OctaMED): Used by the OctaMED software on the Amiga, known for its versatility and support for both sample-based and MIDI instruments.
9. FAR (Farandole Composer): A lesser-known format used by the Farandole Composer tracker, which supported features like pattern looping and volume envelopes.
10. PTM (PolyTracker Module): Created for the PolyTracker software, offering features like extended sample and pattern handling.
11. AMS (Extreme Tracker Module): Used by the Extreme Tracker software, offering unique features and extended sample support.
12. PSM (Protracker Studio Module): Employed by Protracker Studio, which is not to be confused with the original ProTracker on the Amiga.
13. ULT (UltraTracker Module): Associated with UltraTracker, known for its extended pattern commands and sample features.
14. DMF (Delusion Digital Music Format): Used by the XTracker software, supporting features like stereo samples and extended effects.
15. AMF (Dual Module Player / DSMI format)

There are probably others. I haven't used MED, PTM, AMS, PSM, ULT or DMF personally, but I have some files in all the other formats.

It looks likemodplug supports 22 formats, including MOD, S3M, XM, IT, 669, AMF (both of them), AMS, DBM, DMF, DSM, FAR,   MDL, MED, MTM, OKT, PTM, STM, ULT, UMX, MT2, PSM .

IANAL, so I'm not sure if the license is compatible with MA.

Right now, I use a browser-based MOD player and cast my Android device audio. That only allows me to choose a single Chromecast device. My 3 best sound systems use Marantz amps, which don't support Chromecast, but support Airplay and DLNA. A MOD music provider would be much better.

**Comment by imeowfuta** [comment]
Someone else would have to verify, but simply setting --enable-libopenmpt in music assistant's ffmpeg may already cover a lot of formats?
